CRM/Contact/BAO/Contact/Permission.php

```diff
@@ -335,6 +439,104 @@ public static function relationship($selectedContactID, $contactID = NULL) {
+      if (!CRM_Core_Permission::check('access deleted contacts')) {
+        $LEFT_JOIN_DELETED       = "LEFT JOIN civicrm_contact ON civicrm_contact.id = {$contact_id_column} ";
+        $AND_CAN_ACCESS_DELETED  = "AND civicrm_contact.is_deleted = 0";
+      }
+
+      $queries[] = "
+SELECT DISTINCT(civicrm_relationship.{$contact_id_column}) AS contact_id
+  FROM civicrm_relationship 
+  {$LEFT_JOIN_DELETED}
+ WHERE civicrm_relationship.{$user_id_column} = {$contactID}
+   AND civicrm_relationship.{$contact_id_column} IN ({$contact_id_list})
+   AND civicrm_relationship.is_active = 1
+   AND civicrm_relationship.is_permission_{$direction['from']}_{$direction['to']} = 1
+   $AND_CAN_ACCESS_DELETED";
+    }
+
+    if ($config->secondDegRelPermissions) {
+      // ADD SECOND DEGREE RELATIONSHIPS
+      if ($config->secondDegRelPermissions) {
+        foreach ($directions as $first_direction) {
+          foreach ($directions as $second_direction) {
+            // add clause for deleted contacts, if the user doesn't have the permission to access them
+            $LEFT_JOIN_DELETED = $AND_CAN_ACCESS_DELETED = '';
+            if (!CRM_Core_Permission::check('access deleted contacts')) {
+              $LEFT_JOIN_DELETED       = "LEFT JOIN civicrm_contact first_degree_contact  ON first_degree_contact.id  = second_degree_relationship.contact_id_{$second_direction['from']}\n";
+              $LEFT_JOIN_DELETED      .= "LEFT JOIN civicrm_contact second_degree_contact ON second_degree_contact.id = second_degree_relationship.contact_id_{$second_direction['to']} ";
+              $AND_CAN_ACCESS_DELETED  = "AND first_degree_contact.is_deleted = 0\n";
+              $AND_CAN_ACCESS_DELETED .= "AND second_degree_contact.is_deleted = 0 ";
+            }
+
+            $queries[] = "
+  SELECT DISTINCT(second_degree_relationship.contact_id_{$second_direction['to']}) AS contact_id
+    FROM civicrm_relationship first_degree_relationship
+    LEFT JOIN civicrm_relationship second_degree_relationship ON first_degree_relationship.contact_id_{$first_direction['to']} = second_degree_relationship.contact_id_{$first_direction['from']}
+    {$LEFT_JOIN_DELETED}
+   WHERE first_degree_relationship.contact_id_{$first_direction['from']} = {$contactID}
+     AND second_degree_relationship.contact_id_{$second_direction['to']} IN ({$contact_id_list})
+     AND first_degree_relationship.is_active = 1
+     AND first_degree_relationship.is_permission_{$first_direction['from']}_{$first_direction['to']} = 1
+     AND second_degree_relationship.is_active = 1
+     AND second_degree_relationship.is_permission_{$second_direction['from']}_{$second_direction['to']} = 1
+     $AND_CAN_ACCESS_DELETED";
+          }
+        }
+      }
+    }
+
+    // finally UNION the queries and call
+    $query = "(" . implode(")\nUNION (", $queries) . ")";
+    $result = CRM_Core_DAO::executeQuery($query);
+    while ($result->fetch()) {
+      $result_set[(int) $result->contact_id] = TRUE;
```

Comment:
I'd be tempted to skip this loop & do

$query = "SELECT GROUP_CONCAT contact_id FROM ( $query ) as q";
return explode(',', CRM_Core_DAO::singleValueQuery($query));

My guess is more work on mysql & less on php is better